mDNS responder

jukkar opened the pull request:
This allows client like linux to resolve <hostname>.local address that zephyr has.

Todo (but not in this set):
* resolve IP address
* support service resolving

There are patches in this set that are also found in #4085 and #4089

jukkar committed:
net: mdns: Initial mDNS responder support

This creates mDNS responder and serves configured IP addresses
to the callers which want to resolve .local addresses.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: Add hostname setting to Kconfig

User can configure hostname of the device in Kconfig. This can
be used by mDNS responder to answer <hostname>.local queries.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: shell: Print hostname information

When printing network interfaces, print also current hostname
if it has a valid value.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
sample: net: mdns: Add mDNS responder application

This application does not do anything itself, it just waits
mDNS queries and responds to them.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

tbursztyka commented on subsys/net/Kconfig:
```diff
@@ -81,6 +81,29 @@ config  NETWORKING
+	This will append link address to hostname to create a unique
+	hostname. For example, zephyr00005e005357 could be the hostname
+	if this	setting is enabled.
+
```
Maybe creating a Kconfig.hostname would be interesting. 

As well as net_hostname.c, just to move it out of net_core.c which is quite big now.

We may have future additions also like dynamically changing the hostname. (and using such functions through net shell would be interesting as well).

tbursztyka commented on subsys/net/ip/net_core.c:
```diff
@@ -381,6 +428,12 @@ static int net_init(struct device *unused)
+	memcpy(_hostname, CONFIG_NET_HOSTNAME, sizeof(CONFIG_NET_HOSTNAME) - 1);
+
+	NET_DBG("Hostname set to %s", _hostname);
+#endif
```
so here instead it would call net_hostname_init(), a function exposed in net_private.h, something like that.
It will avoid having the ifdef here at least (it will be in net_private.h then).

tbursztyka commented on subsys/net/lib/dns/dns_pack.c:
```diff
@@ -404,3 +406,168 @@ int dns_copy_qname(u8_t *buf, u16_t *len, u16_t size,
+	}
+
+	rc = dns_header_rcode(dns_header);
+	if (rc != 0) {
```
you don't seem to use rc anywhere else, so maybe you can get rid of that variable and test dns_header_rcode() directly.

tbursztyka commented on subsys/net/lib/dns/dns_pack.c:
```diff
@@ -404,3 +406,168 @@ int dns_copy_qname(u8_t *buf, u16_t *len, u16_t size,
+	}
+
+	while ((val = *curr_src++)) {
+		if (val & 0xc0) {
```
would you have a macro name for 0xc0 instead? something that self-describe if possible the value.

tbursztyka commented on subsys/net/lib/dns/dns_pack.c:
```diff
@@ -404,3 +406,168 @@ int dns_copy_qname(u8_t *buf, u16_t *len, u16_t size,
+{
+	const u8_t *curr_src = src;
+	const u8_t *end_of_label = NULL;
+	int dest_size = net_buf_tailroom(buf);
```
I would reverse the order of the 3 lines above, that's a detail though.

tbursztyka commented on subsys/net/lib/dns/dns_pack.c:
```diff
@@ -404,3 +406,168 @@ int dns_copy_qname(u8_t *buf, u16_t *len, u16_t size,
+
+			end_of_label = curr_src + 1;
+
+			pos = ((val & 0x3f) << 8) | (*curr_src & 0xff);
```
same here for 0x3f and 0xff

tbursztyka commented on subsys/net/lib/dns/dns_pack.c:
```diff
@@ -404,3 +406,168 @@ int dns_copy_qname(u8_t *buf, u16_t *len, u16_t size,
+
+			if (((buf->data + label_len + 1) >=
+			     (buf->data + dest_size))
+			    ||
```
weird cut, can you move it at the end of the previous line?

tbursztyka commented on subsys/net/lib/dns/dns_pack.h:
```diff
@@ -328,4 +329,38 @@ int dns_unpack_response_query(struct dns_msg_t *dns_msg);
+ * @param src_id Transaction id, this is returned to the caller.
+ * @retval 0 on success, <0 if error
+ * @retval -ENOMEM if the buffer in msg has no enough space to store the header.
+ * The header is always 12 bytes length.
```
can you align the to previous @<keyword> +1 space, relevantly ?
It won't change anything in doc generation, but in-place reading will be easier. And afaik we did so everywhere.

Apply this on all this patch

jukkar replied:
That could be done in separate patch, all the documentation in this file is now looking uniform.

tbursztyka commented on subsys/net/lib/dns/mdns_responder.c:
```diff
@@ -0,0 +1,520 @@
+	}
+
+	return 0;
+
```
why not falling over what's below quit label? Isn't ret being 0 anyway when everything went well?

tbursztyka commented on subsys/net/lib/dns/mdns_responder.c:
```diff
@@ -0,0 +1,520 @@
+		goto quit;
+	}
+
+quit:
```
looks like quit label should be put at line 404, all the relevant gotos come from places where dns_data is not set.

jukkar committed:
net: dns: Fix dns_pack.h documentation of packing routines

Some doxygen directives were missing from dns_pack.h file.
Also make function header documentation look better.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: ipv4: Accept multicast packets for processing

We dropped received IPv4 multicast packet.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

jukkar committed:
net: ethernet: Check multicast IPv4 MAC address

Make sure the multicast MAC address checker checks also
IPv4 multicast MAC address and accepts it.

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

pfalcon commented:
> There are patches in this set that are also found in #4085 and #4089

Should this be rebased then?

jukkar commented:
> Should this be rebased then?

No, those are already merged. Why would this be rebased anyway?

pfalcon commented:
Because this pull request lists 12 commits, some of which are apparently merged. (So, to get rid of them and decrease the complexity of the review, it needs to be rebased.)

jukkar commented:
@tbursztyka Are you ok with the changes?

pfalcon commented on subsys/net/lib/dns/mdns_responder.c:
```diff
@@ -0,0 +1,520 @@
+/** @file
+ * @brief mDNS responder
+ *
+ * This listenes mDNS queries and responds to them.
```
Typo: s/listenes/listens to/.

jukkar commented:
New version that fixes the typo in mdns_responder.c

tbursztyka reviewed (commented):
Unlocking the PR, most of it looks ok.

jukkar commented:
I will send more changes to this that address the hostname issues.

jukkar commented:
Updating the PR with correct fixes.

jukkar commented:
Adding @nashif to review list, I hope you have some time to review this.
